plugins/osdn/netid/controller/controller.go

```diff
@@ -0,0 +1,232 @@
+package controller
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+	kcache "k8s.io/kubernetes/pkg/client/cache"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+	utilwait "k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/openshift/openshift-sdn/pkg/netid"
+	"github.com/openshift/openshift-sdn/plugins/osdn/netid/vnidallocator"
+	"github.com/openshift/origin/pkg/client/cache"
+	"github.com/openshift/origin/pkg/controller"
+)
+
+type VnidController struct {
+	vnid   vnidallocator.Interface
+	client kclient.NamespaceInterface
+	Queue  *cache.EventQueue
+
+	// globalNamespaces are the namespaces that have access to all pods in the cluster and vice versa.
+	globalNamespaces sets.String
+}
+
+func NewVnidController(vnida vnidallocator.Interface, client kclient.NamespaceInterface, globalNamespaces []string) *VnidController {
+	return &VnidController{
+		vnid:             vnida,
+		client:           client,
+		globalNamespaces: sets.NewString(globalNamespaces...),
+	}
+}
+
+// Create creates a vnid controller
+func (c *VnidController) Create() controller.RunnableController {
+	if c.Queue == nil {
+		lw := &kcache.ListWatch{
+			ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
+				return c.client.List(options)
+			},
+			WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
+				return c.client.Watch(options)
+			},
+		}
+		q := cache.NewEventQueue(kcache.MetaNamespaceKeyFunc)
+		kcache.NewReflector(lw, &kapi.Namespace{}, q, 10*time.Minute).Run()
+		c.Queue = q
+	}
+	return c
+}
+
+// Run begins processing resources from Queue asynchronously.
+func (c *VnidController) Run() {
+	go utilwait.Forever(func() { c.handleOne(c.Queue.Pop()) }, 0)
+}
+
+// handleOne processes resource with Handle.
+func (c *VnidController) handleOne(eventType watch.EventType, resource interface{}, err error) {
+	if err != nil {
+		glog.Errorf("Failed to pop namespace from event queue: %v", err)
+		return
+	}
+
+	ns := resource.(*kapi.Namespace)
+	switch eventType {
+	case watch.Added, watch.Modified:
+		err = c.addOrUpdate(ns)
+	case watch.Deleted:
+		err = c.delete(ns)
+	}
+	if err != nil {
+		glog.Errorf("Failed to apply vnid allocation, event-type: %v, namespace: %v, error: %v", eventType, ns, err)
+	}
+}
+
+// retryCount is the number of times to retry on a conflict when updating a namespace
+const retryCount = 2
+
+// addOrUpdate allocates network id(VNID) for the namespace if needed
+func (c *VnidController) addOrUpdate(ns *kapi.Namespace) error {
```

Comment:
Does this controller go away with Jordan's suggested changes for tolerating old and new?